### ISSUE TITLE
Does early stopping callback ensure the best model parameters are stored?

### ISSUE THREAD

**Issue by pdasigi** [fchollet/keras#2611]
It looks like the callback causes training to stop if the validation loss increases, which means the parameters from the last epoch are the ones that are kept. Can someone confirm this? Should it not be the case that we store the parameters from the epoch before the last one instead, which result in the lowest validation loss?

**Comment by tboquet** [comment]
Sure you could combine `ModelCheckpoint` using `save_best_only=False` and a fixed path (so that you end up with the last weights) and an `EarlyStopping` callback. You could take a look at the source of the `ModelCheckpoint` callback and modify it if you want to keep the weights in RAM and skip the hdf5 dump part.

**Comment by pdasigi** [comment]
Thanks for the suggestion @tboquet ! I'll look into it. But should't storing the best model be the default behavior of EarlyStopping?

**Comment by tboquet** [comment]
The weights live in the GPU VRAM or the RAM and if you have a big model, you will end up having the weights twice in memory: the current weights updated in the network and the weights corresponding to your best results. Callbacks are easier to customize and everyone has a different problem, hardware or dataset. It's a good tradeoff between simplicity and customization since you could already do what you want using built-in callbacks.

**Comment by pdasigi** [comment]
I agree. Closing the issue.

**Comment by damnko** [comment]
Hi, I have a similar question related to `ModelCheckpoint` and `EarlyStopping`.

Let's suppose I'm fitting a model and using both the above callbacks, when the training finishes, the `model` instance will have stored the best model or should I actively load the best weights stored by `ModelCheckpoint` before using `model.evaluate(..)` or doing any prediction?

Thanks
